model/errors/transaction.go

```diff
@@ -0,0 +1,459 @@
+// this error is the result of failure in any of the following conditions:
+// - keyIndex doesn't exist at this address
+type EnvelopeSignatureKeyError struct {
+	TxHash   flow.Identifier
+	Address  flow.Address
+	KeyIndex uint64
+}
+
+func (e *EnvelopeSignatureKeyError) TxHash() flow.Identifier {
+	return e.TxHash
+}
+
+func (e *EnvelopeSignatureKeyError) Code() uint32 {
+	return errCodeEnvelopeSignatureKeyError
+}
+
+func (e *EnvelopeSignatureKeyError) Error() string {
+	return fmt.Sprintf(
+		"invalid envelope key: key index %d doesn't exist on account %s",
+		e.KeyIndex,
+		e.Address,
+	)
+}
+
+// RevokedEnvelopeSignatureKeyError indicates a transaction payload key is revoked.
+// this error is the result of failure in any of the following conditions:
+// - key Index is revoked from this account
+// TODO maybe merge with the one above
+type RevokedEnvelopeSignatureKeyError struct {
+	TxHash   flow.Identifier
+	Address  flow.Address
+	KeyIndex uint64
+}
+
+func (e *RevokedEnvelopeSignatureKeyError) TxHash() flow.Identifier {
+	return e.TxHash
+}
+
+func (e *RevokedEnvelopeSignatureKeyError) Code() uint32 {
+	return errCodeRevokedEnvelopeSignatureKeyError
+}
+
+func (e *RevokedEnvelopeSignatureKeyError) Error() string {
+	return fmt.Sprintf(
+		"invalid envelope key: key index %d doesn't exist on account %s",
+		e.KeyIndex,
+		e.Address,
+	)
+}
+
+// AuthorizationError indicates that a transaction is missing a required signature to
+// authorize access to an account.
+// this error is the result of failure in any of the following conditions:
+// - no signature provided for an account
+// - not enough key weight in total for this account
+type AuthorizationError struct {
+	TxHash       flow.Identifier
+	Address      flow.Address
+	SignedWeight uint32
+}
+
+func (e *AuthorizationError) TxHash() flow.Identifier {
+	return e.TxHash
+}
+
+func (e *AuthorizationError) Code() uint32 {
+	return errCodeAuthorizationError
+}
+
+func (e *AuthorizationError) Error() string {
+	return fmt.Sprintf(
+		"account %s does not have sufficient signatures (unauthorized access)",
+		e.KeyIndex,
+		e.Address,
+	)
+}
+
+// CadenceError captures a collection of errors provided by cadence runtime
+// it cover cadence errors such as
+// NotDeclaredError, NotInvokableError, ArgumentCountError, TransactionNotDeclaredError,
+// ConditionError, RedeclarationError, DereferenceError,
+// OverflowError, UnderflowError, DivisionByZeroError,
+// DestroyedCompositeError,  ForceAssignmentToNonNilResourceError, ForceNilError,
+// TypeMismatchError, InvalidPathDomainError, OverwriteError, CyclicLinkError,
+// ArrayIndexOutOfBoundsError, ...
+type CadenceRunTimeError struct {
+	TxHash flow.Identifier
+	error  runtime.Error
+}
+
+// TODO add unwrap
+
+// account doesn't have enough storage
+type InsufficientStorageError struct {
+	Capacity uint64
+	Used     uint64
+	Needed   uint64
+}
+
+func (e InsufficientStorageError) TxHash() flow.Identifier {
+	return e.TxHash
+}
```

Comment:
The only way to actually call these `TxHash` methods would be to type assert to the structure type (ie. `InsufficientStorageError`) in which case you can just do `err.TxHash`. I think this scheme causes naming collisions as well (`err.TxHash` is ambiguous currently)